llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp

```diff
@@ -1337,7 +1339,17 @@ uint32_t DWARFDebugLine::LineTable::lookupAddressImpl(
                                       DWARFDebugLine::Sequence::orderByHighPC);
   if (It == Sequences.end() || It->SectionIndex != Address.SectionIndex)
     return UnknownRowIndex;
-  return findRowInSeq(*It, Address);
+
+  uint32_t RowIndex = findRowInSeq(*It, Address);
+  if (RowIndex != UnknownRowIndex && DWARFDebugLine::ReportApproximateLine) {
+    for (uint32_t ApproxRowIndex = RowIndex;
+         ApproxRowIndex >= It->FirstRowIndex; --ApproxRowIndex) {
+      if (Rows[ApproxRowIndex].Line)
+        return ApproxRowIndex;
+      *IsApproximateLine = true;
+    }
+  }
+  return RowIndex;
```

Comment:
I think it might be clearer to have the early return, rather than having the non-approximate case come out through this for loop?

```
if (RowIndex != UnknownRowIndex || !ReportApproximateLine)
  return RowIndex;
for (...) {
  if (Rows[ApproxRowIndex].Line) {
    *IsApproximateLine = true;
    return ApproxRowIndex;
  }
}
return RowIndex;
```

Also, this will ensure that `(approximate)` isn't printed if the approximation doesn't find a non-zero line number (might be worth adding a test for that to demonstrate the fix/change in behavior here).

Reply:
You'd probably need a hand-crafted line table (the .line directives in the assembly - it's probably easier to build something from scratch rather than try to modify the assembly from some real-world code)

A simple .text section, with a line directive, of, say line 0 with file A, then a nop instruction, then another line directive of line 0 with file B.

Symbolize the second address (should be literally symbolizing the value 1) and check that without the code to ensure that we return the original row where the address was found, we'd end up returning the 0 line at the start of the sequence and that would mean the symbolized output would mention file A. But with the code to ensure that, failing to find a non-zero line in the sequence, we prefer the original - we should see file B mentioned, not file A.

(might be worth introducing more than one sequence, to check that we don't walk into another sequence - and probably clearer/simpler testing it with a little hand-crafted assembly like this rather than full blown generated-from-the-frontend assembly)

Reply:
I have added a new file ```approximate-line-handcrafted.s``` assembly to cover this functionality.

The ```approximate-line-handcrafted.s``` is generated from clang, however I have manually edited the line entries(.loc) at some position in the assembly file to cover that if approximation loop fails then it should return the original rowindex.

I also created two sequences to achieve this. It is hard to create sequences manually in assembly.  Also, the output object file  generated from the assembly still have one sequence always no matter whatever I did.  To overcome this I had switched to usage of custom sections. Custom sections enforce sequence boundaries in the object file.  Hence the approximate-line-handcrafted.s covers two sequences 

One sequence  is for main.c file
Another sequence is for definitions.h file